lib/rack/utils.rb

```diff
@@ -607,9 +607,7 @@ def clean_path_info(path_info)
         part == '..' ? clean.pop : clean << part
       end
 
-      clean_path = clean.join(::File::SEPARATOR)
-      clean_path.prepend("/") if parts.empty? || parts.first.empty?
-      clean_path
+      "#{::File::SEPARATOR}#{clean.join(::File::SEPARATOR)}"
```

Comment:
This results in different behavior for relative paths (e.g. `"a"`, `"b/a"`).  Previously, `clean_path_info` did not prepend a slash, and now it would. For that reason, I doubt it would be acceptable.

Looking at the history, it appears to be called with `PATH_INFO`, which generally should be empty or start with a slash.  However, the specs have tests with relative paths.  Those tests happen to work with this, but that seems like an accident and not deliberate.  For example, there is a test for `"test/.."`, but not a test for `"test"` by itself.

As we cannot be sure how users are using `clean_path_info`, I don't think we should accept the behavior change.

Reply:
Could you say in what case `PATH_INFO` would be empty or relative? What valid HTTP request will make it empty?

I think it is important for this method to be as safe as possible for the intended purpose - to make the PATH_INFO predictable to compare or use regular expression on. Rather than some unintended use cases.

Presently just figuring out what would be the actual path of a request is rather hard. And it should not be hard to do it right.

This means, as I mentioned to prevent https://owasp.org/www-community/attacks/Path_Traversal

Reply:
HTTP 1.1 supports only (https://datatracker.ietf.org/doc/html/rfc9112#name-request-target) 
* absolute path
* full URL (in which case server sets PATH_INFO to an absolute path)
* host:port (in CONNECT requests but we are not a proxy)
* `*` (single star for server global OPTIONS request, but we can handle this as an individual check)

So basically I see no valid use cases where the result should be a relative path.

Reply:
Just to clarify, we do check those four scenarios here: https://github.com/rack/rack/blob/fc541f3f443ed170b9b1daed4d16c6407a12573d/lib/rack/lint.rb#L364-L389

Reply:
@ioquatix , I wonder why empty string is accepted. RFC clearly reads:
>  If the target URI's path component is empty, the client MUST send "/" as the path within the origin-form of request-target.

Reply:
Additionally the RFC states:
> Recipients of an invalid request-line SHOULD respond with either a 400 (Bad Request) error or a 301 (Moved Permanently) redirect with the request-target properly encoded. A recipient SHOULD NOT attempt to autocorrect and then process the request without a redirect, since the invalid request-line might be deliberately crafted to bypass security filters along the request chain.

Which is to a big extend my concern.